### ISSUE TITLE
Hello, I saw `ibhp` (Indian Buffet Hawkes Process) in one of the source files in this repository, have you ever reproduced this model? If so, can I have a discussion about this model with you?

### ISSUE THREAD

**Issue by hjnnjh** [gauenk/hawkes_process#1]

**Comment by gauenk** [comment]
I have not implemented the method. I have been considering implementing it as an exercise since I can't find a github of it online. What questions do you have about it? I can try my best to answer.

**Comment by hjnnjh** [comment]
> I have not implemented the method. I have been considering implementing it as an exercise since I can't find a github of it online. What questions do you have about it? I can try my best to answer.

Thanks for your reply! I'm trying to implement this model. There are many problems in my implementation, so I set my repository as a private one that can not be searched.  If you are interested in it, I can send you an invitation.

Here are my questions about IBHP:

- From the point of view of factors, this model looks like a multidimensional Hawkes process with a sum exponential kernel, but from the point of view of events, it seems like a univariate Hawkes process. I generated simulation data according to Hawkes process simulation algorithm (by thinning, shown in the following picture). I am not sure whether this algorithm can be used for the simulation data generation of this model, which the author did not mention in the paper, but he implemented an experiment based on simulation data.
![image](https://user-images.githubusercontent.com/37699150/160285710-b737c747-3a97-46e2-8498-2fa02fc040c7.png)

- I am not very clear about the step of `Sample Model Parameters` in SMC algorithm. The author mentioned that he used HP likelihood（Hawkes Process likelihood）. However, for IBHP, this likelihood seems to be very complicated. I am not very clear about how the author calculates this likelihood and which algorithm is adopted to sample Model Parameters.

- In the simulation data experiment, the author only set up 10 particles to achieve a good model effect, is this really possible?

Currently, I have a study on the basis of IBHP, I would really appreciate it if you could help me.

**Comment by gauenk** [comment]
Thank you for your detailed message :D I will try my best to answer each part below:

> - "From the point of view of factors, this model looks like a multidimensional Hawkes process with a sum exponential kernel, but from the point of view of events, it seems like a univariate Hawkes process." 

- I am not sure what is meant by  "multidimensional" v.s. a "univariate" Hawkes process. Maybe you are talking about how the rate of the process in Equation 12 is a sum of events ([paper here](http://auai.org/uai2018/proceedings/papers/289.pdf)). If this is true, then I see what you mean since the final rate is a sum of *multiple* rates.

> - "I generated simulation data according to Hawkes process simulation algorithm (by thinning, shown in the following picture). I am not sure whether this algorithm can be used for the simulation data generation of this model, which the author did not mention in the paper, but he implemented an experiment based on simulation data."

- From what I understand, there are two major categories of methods to sample ~exactly~ from a generative model. One set of methods *approximately* sample from a process (named Approximate Bayesian Computation). Another set of methods *exactly* sample from a process ([see MacKay's chapter here](http://www.inference.org.uk/mackay/itprnn/ps/413.435.pdf)). The "thinning" procedure you are using I think comes form an exact method, while the authors use Sequential Monte Carlo (SMC), an approximate method.
- For the record, I think SMC is considered an always-available/simple way to sample from a model. So if you write down something that looks complication (like this in this IBHP paper) then SMC can be used to ~at least get some samples from your prior~ get samples from the posterior as long you can generate sample from the prior (I think)

> - I am not very clear about the step of Sample Model Parameters in SMC algorithm. The author mentioned that he used HP likelihood（Hawkes Process likelihood）. However, for IBHP, this likelihood seems to be very complicated. I am not very clear about how the author calculates this likelihood and which algorithm is adopted to sample Model Parameters.

- I am also confused on this too right now. I see on page 4 the authors cite reference 8, so maybe they are just using the likelihood in that reference... Looking at that [reference here](https://dl.acm.org/doi/pdf/10.1145/2783258.2783411), maybe its Equation 21?

> - In the simulation data experiment, the author only set up 10 particles to achieve a good model effect, is this really possible?

- The results in Table 1 suggest 10 particles was enough for their experiment.

**Comment by hjnnjh** [comment]
Thank you very much for your reply!
> From what I understand, there are two major categories of methods to sample exactly from a generative model. One set of methods approximately sample from a process (named Approximate Bayesian Computation). Another set of methods exactly sample from a process ([see MacKay's chapter here](http://www.inference.org.uk/mackay/itprnn/ps/413.435.pdf)). The "thinning" procedure you are using I think comes form an exact method, while the authors use Sequential Monte Carlo (SMC), an approximate method.

Maybe you misunderstood me, What I mean is that the author generates 1000 observations, a.k.a. 1000 events, as training data and test data respectively for training the IBHP model and evaluating the model performance.
![image](https://user-images.githubusercontent.com/37699150/160631201-72885fc9-77a0-4a13-b30f-44d049fd62d2.png)
What I'm talking about here is what method the author used to generate these 1000 observations, and these observations include timestamps and text. 

> I am also confused on this too right now. I see on page 4 the authors cite reference 8, so maybe they are just using the likelihood in that reference... Looking at that [reference here](https://dl.acm.org/doi/pdf/10.1145/2783258.2783411), maybe its Equation 21?

I derived the expression of likelihood function according to the intensity function of IBHP model, but when I verified its correctness, I found that the HP likelihood calculated using real parameters was not the maximum. Is this reasonable in Hawkes Process?

**Comment by gauenk** [comment]
Ah okay - I see your question now. I think they are sampling from the prior and using the samples as data to estimate parameters of the posterior for two reasons: (1) to understand the identifiability of our model and the accuracy of our SMC algorithm when the true data generation process satisfies the model assumptions, and (2) to understand the effects of misspecification. Does this help?

TLDR; The authors take 1000 exact samples from the prior.
